WalletWasabi/Crypto/ZeroKnowledge/Verifier.cs

```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge
+{
+	public static class Verifier
+	{
+		public static bool Verify(KnowledgeOfRepresentation proof, GroupElement publicPoint, IEnumerable<GroupElement> generators)
+		{
+			Guard.False($"{nameof(publicPoint)}.{nameof(publicPoint.IsInfinity)}", publicPoint.IsInfinity);
+			Guard.NotNullOrEmpty(nameof(generators), generators);
+
+			if (publicPoint == proof.Nonce)
+			{
+				throw new InvalidOperationException($"{nameof(publicPoint)} and {nameof(proof.Nonce)} should not be equal.");
+			}
+
+			foreach (var generator in generators)
+			{
+				Guard.False($"{nameof(generator)}.{nameof(generator.IsInfinity)}", generator.IsInfinity);
+			}
+
+			var nonce = proof.Nonce;
+			var responses = proof.Responses;
+
+			var challenge = Challenge.Build(publicPoint, nonce, generators);
+			var a = challenge * publicPoint + nonce;
+
+			var b = GroupElement.Infinity;
+			foreach (var (response, generator) in responses.ZipForceEqualLength(generators))
```

Comment:
IMO it should be better to check at the begining of the method that there is a `response` for each `generator`.

Reply:
I think the software crashing by mistakes is a worse idea than not even letting making those mistakes in the first place. I agree it looks a bit clunky, but it's more secure and after all isn't that what really matters in crypto code?